pkg/contract/contract.go

```diff
@@ -1,225 +1,121 @@
-	parsedTime, err := monday.ParseInLocation(timeLayout, timeStr, amsterdamLocation, monday.LocaleNlNL)
-	if err != nil {
-		return nil, fmt.Errorf("invalid time string [%v]: %w", timeStr, err)
+	sc.Params = make(map[string]string, len(matches))
+	for idx, match := range matches {
+		sc.Params[sc.Template.TemplateAttributes[idx]] = string(match)
 	}
-	return &parsedTime, nil
+	return nil
 }
 
-func (c Template) ValidateTimeFrame(params map[string]string) (bool, error) {
+// Verify verifies the params with the template
+func (sc Contract) Verify() error {
 	var (
 		err                      error
 		ok                       bool
 		validFrom, validTo       *time.Time
 		validFromStr, validToStr string
 	)
 
-	if validFromStr, ok = params["valid_from"]; !ok {
-		return false, fmt.Errorf("%w: value for [valid_from] is missing", ErrInvalidContractText)
+	if validFromStr, ok = sc.Params[ValidFromAttr]; !ok {
+		return fmt.Errorf("%w: value for ["+ValidFromAttr+"] is missing", ErrInvalidContractText)
 	}
 
-	validFrom, err = parseTime(validFromStr, c.Language)
+	validFrom, err = parseTime(validFromStr, sc.Template.Language)
 	if err != nil {
-		return false, fmt.Errorf("%w: unable to parse [valid_from]: %s", ErrInvalidContractText, err)
+		return fmt.Errorf("%w: unable to parse ["+ValidFromAttr+"]: %s", ErrInvalidContractText, err)
 
 	}
 
-	if validToStr, ok = params["valid_to"]; !ok {
-		return false, fmt.Errorf("%w: value for [valid_to] is missing", ErrInvalidContractText)
+	if validToStr, ok = sc.Params[ValidToAttr]; !ok {
+		return fmt.Errorf("%w: value for ["+ValidToAttr+"] is missing", ErrInvalidContractText)
 	}
 
-	validTo, err = parseTime(validToStr, c.Language)
+	validTo, err = parseTime(validToStr, sc.Template.Language)
 	if err != nil {
-		return false, fmt.Errorf("%w: unable to parse [valid_to]: %s", ErrInvalidContractText, err)
+		return fmt.Errorf("%w: unable to parse ["+ValidToAttr+"]: %s", ErrInvalidContractText, err)
 	}
 
 	// All parsed, check time range
 	if validFrom.After(*validTo) {
-		return false, fmt.Errorf("%w: [valid_from] must be after [valid_to]", ErrInvalidContractText)
+		return fmt.Errorf("%w: ["+ValidFromAttr+"] must be after ["+ValidToAttr+"]", ErrInvalidContractText)
 	}
 
 	amsterdamLocation, _ := time.LoadLocation("Europe/Amsterdam")
 	now := NowFunc()
 	logrus.Debugf("checking timeframe: now %v, validFrom: %v, validTo: %v", now, *validFrom, *validTo)
 
 	if now.In(amsterdamLocation).Before(*validFrom) {
-		logrus.Infof("contract is not yet valid. now: %s, validFrom: %s", now, validFrom)
-		return false, nil
+		return fmt.Errorf("contract is not yet valid. now: %s, validFrom: %s", now, validFrom)
 
 	}
 	if now.In(amsterdamLocation).After(*validTo) {
-		logrus.Info("contract is expired")
-		return false, nil
+		return fmt.Errorf("contract is expired since: %s", validTo)
 	}
 
-	return true, nil
+	return nil
 }
 
-type Matrix map[Language]map[Type]map[Version]*Template
+const amsterdamTimeZone = "Europe/Amsterdam"
```

Comment:
Maybe also add a todo for using OS timezone

Reply:
Would that work on a machine hosted on a foreign cloud?

Reply:
a quick search results in issues indeed. This can also be annoying when using a cloud hosted DB with datetime fields. A todo for configuring the timezone then ;)